packages/server/src/scim.ts

```diff
@@ -0,0 +1,126 @@
+import { DefaultScimProvider, ScimConfig, ScimUserRequestData } from "@padloc/core/src/scim";
+import { MemberProvisioning, Provisioner } from "@padloc/core/src/provisioning";
+
+import { createServer, IncomingMessage, ServerResponse } from "http";
+import { readBody } from "./transport/http";
+
+export class ScimProvider extends DefaultScimProvider {
+    constructor(public readonly config: ScimConfig, public readonly provisioner: Provisioner) {
+        super(config, provisioner);
+    }
+
+    async init() {
+        await this._startScimServer();
+    }
+
+    private _getDataFromScimRequest(httpReq: IncomingMessage) {
+        const url = new URL(`http://localhost${httpReq.url || ""}`);
+        const secretToken = url.searchParams.get("token") || "";
+        const orgId = url.searchParams.get("org") || "";
+
+        // TODO: find account/org based on token
+        if (secretToken === "asdrtyghj") {
+            return {
+                accountId: "472478c5-17e8-4ed5-8f51-05a9c5deaebb",
+                orgId,
+            };
+        }
+
+        return {
+            accountId: null,
+            orgId: null,
+        };
+    }
+
+    private async _handleScimUsersPost(httpReq: IncomingMessage, httpRes: ServerResponse) {
+        let newUser: ScimUserRequestData;
+
+        const { accountId, orgId } = this._getDataFromScimRequest(httpReq);
+
+        if (!accountId || !orgId) {
+            httpRes.statusCode = 400;
+            httpRes.end("Invalid SCIM Secret Token");
+            return;
+        }
+
+        try {
+            const body = await readBody(httpReq);
+            newUser = JSON.parse(body);
+        } catch (e) {
+            httpRes.statusCode = 400;
+            httpRes.end("Failed to read request body.");
+            return;
+        }
+
+        const validationError = this.validateScimUser(newUser);
+        if (validationError) {
+            httpRes.statusCode = 400;
+            httpRes.end(validationError);
+            return;
+        }
+
+        try {
+            const provisioning = await this.provisioner.getProvisioning({ email: newUser.email });
+            const orgProvisioning = provisioning.orgs.find((org) => org.id === orgId);
+
+            if (!orgProvisioning) {
+                throw new Error("Organization not found");
+            }
+
+            const newProvisioningMember = new MemberProvisioning();
+            newProvisioningMember.email = newUser.email;
+
+            orgProvisioning.members.push(newProvisioningMember);
+
+            // TODO: Save?
+
+            // TODO: create auth provisioning
+            // TODO: create invite provisioning
```

Comment:
@MaKleSoft I believe these are superseded by the `MemberProvisioning` above, correct?

Reply:
The "invite provisioning" yes, but you'll still have to get or create the AccountProvisioning object for that email and make sure the `ProvisioningStatus` is set to `ProvisioningStatus.Active`.

Reply:
In that regard it kind of overlaps with `BasicProvisioner`, which is why I think it should be a subclass/extension.

Reply:
In that regard it kind of overlaps with `BasicProvisioner`, which is why I think it should be a subclass/extension.

Reply:
So specifically:

```sh
const accountProv = await this._getOrCreateAccountProvisioning({ email });
accountProv.status = newUser.active ? ProvisioningStatus.Active : ProvisioningStatus.Suspended;
```

Reply:
👆This is the part that clashes with `StripeProvisioner`, basically.